File: DSA/priorityQueue.js


class PriorityQueue{
    constructor(){
        this.heap=[]
    }

     getLeftChildIndex(parentIndex){
        return parentIndex*2+1}
     getRightChildIndex(parentIndex){
        return parentIndex*2+2}
     getParentIndex(childIndex){
        return Math.ceil(childIndex/2)-1}
     hasLeftChild(parentIndex){
        return this.getLeftChildIndex(parentIndex)<this.heap.length}
     hasRightChild(parentIndex){
        return this.getRightChildIndex(parentIndex)<this.heap.length}
     hasParent(childIndex){
        return this.getParentIndex(childIndex)>=0}
     getLeftChild(parentIndex){
        return this.heap[this.getLeftChildIndex(parentIndex)]}
     getRightChild(parentIndex){
        return this.heap[this.getRightChildIndex(parentIndex)]}
     getParent(childIndex){
        return this.heap[this.getParentIndex(childIndex)]}
     swap(frmIndex,toIndex){
        let temp=this.heap[frmIndex]
        this.heap[frmIndex]=this.heap[toIndex]
        this.heap[toIndex]=temp}
     peek(){
        if (this.heap.length==0){
            return null
        }
        return this.heap[0]
    }
     maxHeapifyUp(){
        let cidx=this.heap.length-1
        while (this.hasParent(cidx) && this.getParent(cidx)<this.heap[cidx]){
            let pIndx=this.getParentIndex(cidx)
            this.swap(cidx,pIndx)
            cidx=pIndx
        }
        }
     maxHeapifyDown(){
        let pidx=0
        while (true){
            if (this.hasRightChild(pidx) && this.getRightChild(pidx)>this.heap[pidx]){
                let cidx=this.getRightChildIndex(pidx)
                this.swap(pidx,cidx)
                pidx=cidx
            }
            else if (this.hasLeftChild(pidx) && this.getLeftChild(pidx)>this.heap[pidx]){
                let cidx=this.getLeftChildIndex(pidx)
                this.swap(pidx,cidx)
                pidx=cidx
            }
            else{
                break
            }
        }
    }
     minHeapifyUp(){
        let cidx=this.heap.length-1
        while (this.hasParent(cidx) && this.getParent(cidx)>this.heap[cidx]){
            let pIndx=this.getParentIndex(cidx)
            this.swap(cidx,pIndx)
            cidx=pIndx
        }
    }
     minHeapifyDown(){
        let pidx=0
        while (true){
            if (this.hasRightChild(pidx) && this.getRightChild(pidx)<this.heap[pidx]){
                let cidx=this.getRightChildIndex(pidx)
                this.swap(pidx,cidx)
                pidx=cidx
            }
            else if (this.hasLeftChild(pidx) && this.getLeftChild(pidx)<this.heap[pidx]){
                let cidx=this.getLeftChildIndex(pidx)
                this.swap(pidx,cidx)
                pidx=cidx
            }
            else{
                break
            }
        }
    }
            
     add(value){
        this.heap.push(value)
        // this.maxHeapifyUp()
        this.minHeapifyUp()
    }
    
     remove(){
        if (this.heap.length==0){
            return null}
        let removed=this.heap[0]
        this.heap[0]=this.heap[this.heap.length-1]
        this.heap.pop()
        //  this.maxHeapifyDown()
        this.minHeapifyDown()
        return removed
        }
    
}


PriQueue = new PriorityQueue()
 
// # Adding the Elements
PriQueue.add(32)
PriQueue.add(45)
PriQueue.add(12)
PriQueue.add(65)
PriQueue.add(85)
 
console.log(PriQueue.heap)
console.log('peek',PriQueue.peek())
console.log('removed',PriQueue.remove())
console.log('peek',PriQueue.peek())
console.log('removed',PriQueue.remove())
console.log('peek',PriQueue.peek())
console.log('removed',PriQueue.remove())
console.log(PriQueue.heap)

File: DSA/priorityQueue.py

import math


class PriorityQueue:
    def __init__(self):
        self.heap=[]
    
    def getLeftChildIndex(self,parentIndex):
        return parentIndex*2+1
    def getRightChildIndex(self,parentIndex):
        return parentIndex*2+2
    def getParentIndex(self,childIndex):
        return math.ceil(childIndex/2)-1
    def hasLeftChild(self,parentIndex):
        return self.getLeftChildIndex(parentIndex)<len(self.heap)
    def hasRightChild(self,parentIndex):
        return self.getRightChildIndex(parentIndex)<len(self.heap)
    def hasParent(self,childIndex):
        return self.getParentIndex(childIndex)>=0
    def getLeftChild(self,parentIndex):
        return self.heap[self.getLeftChildIndex(parentIndex)]
    def getRightChild(self,parentIndex):
        return self.heap[self.getRightChildIndex(parentIndex)]
    def getParent(self,childIndex):
        return self.heap[self.getParentIndex(childIndex)]
    def swap(self,frmIndex,toIndex):
        temp=self.heap[frmIndex]
        self.heap[frmIndex]=self.heap[toIndex]
        self.heap[toIndex]=temp
    def peek(self):
        if len(self.heap)==0:
            return None
        return self.heap[0]
    def maxHeapifyUp(self):
        cidx=len(self.heap)-1
        while self.hasParent(cidx) and self.getParent(cidx)<self.heap[cidx]:
            pIndx=self.getParentIndex(cidx)
            self.swap(cidx,pIndx)
            cidx=pIndx
    def maxHeapifyDown(self):
        pidx=0
        while True:
            if self.hasRightChild(pidx) and self.getRightChild(pidx)>self.heap[pidx]:
                cidx=self.getRightChildIndex(pidx)
                self.swap(pidx,cidx)
                pidx=cidx
            elif self.hasLeftChild(pidx) and self.getLeftChild(pidx)>self.heap[pidx]:
                cidx=self.getLeftChildIndex(pidx)
                self.swap(pidx,cidx)
                pidx=cidx
            else:
                break
    def minHeapifyUp(self):
        cidx=len(self.heap)-1
        while self.hasParent(cidx) and self.getParent(cidx)>self.heap[cidx]:
            pIndx=self.getParentIndex(cidx)
            self.swap(cidx,pIndx)
            cidx=pIndx
    def minHeapifyDown(self):
        pidx=0
        while True:
            if self.hasRightChild(pidx) and self.getRightChild(pidx)<self.heap[pidx]:
                cidx=self.getRightChildIndex(pidx)
                self.swap(pidx,cidx)
                pidx=cidx
            elif self.hasLeftChild(pidx) and self.getLeftChild(pidx)<self.heap[pidx]:
                cidx=self.getLeftChildIndex(pidx)
                self.swap(pidx,cidx)
                pidx=cidx
            else:
                break
            
    def add(self,value):
        self.heap.append(value)
        # self.maxHeapifyUp()
        self.minHeapifyUp()
    
    def remove(self):
        if len(self.heap)==0:
            return None
        removed=self.heap[0]
        self.heap[0]=self.heap[len(self.heap)-1]
        self.heap.pop()
        # self.maxHeapifyDown()
        self.minHeapifyDown()
        return removed
    
    

PriQueue = PriorityQueue()
 
# Adding the Elements
PriQueue.add(32)
PriQueue.add(45)
PriQueue.add(12)
PriQueue.add(65)
PriQueue.add(85)
 
print(PriQueue.heap)
print('peek',PriQueue.peek())
print('removed',PriQueue.remove())
print('peek',PriQueue.peek())
print('removed',PriQueue.remove())
print('peek',PriQueue.peek())
print('removed',PriQueue.remove())
print(PriQueue.heap)